### PR TITLE
feat: Publish to npm registry

### PR DESCRIPTION
**Pull request by ianlewis**
Updates #1669 

Adds support for publishing to the npm registry using a [fork of `npm/cli`](https://github.com/ianlewis/cli) that supports publishing provenance that was generated separately from `npm` itself.

Note that this PR also modifies the BYOB `buildInvocationId` for SLSA v0.2 to `<GITHUB_RUN_ID>-<GITHUB_RUN_ATTEMPT>` because npm seems to validate it.

Example successful published package: https://www.npmjs.com/package/@ianlewis/actions-test/v/0.1.72
Example successful GHA run: https://github.com/ianlewis/actions-test/actions/runs/4486596545/attempts/1

### CODE REVIEW

File: internal/nodejs-action/generate-layout.sh
Comment (laurentsimon) [inline internal/nodejs-action/generate-layout.sh:18]: Given the examples above, it seems `raw_package_name` would never be empty?
`pkg:npm/<scope>/<name>@<version>` -> `<scope>`
`pkg:npm/<name>@<version>` -> `<name>@<version>`

File: .github/workflows/builder_nodejs_slsa3.yml
Comment (laurentsimon) [inline .github/workflows/builder_nodejs_slsa3.yml:138]: I think all these env variables are untrusted (either provided by the TRW or the PW), so pre-pending `UNTRUSTED_` to all of us would be preferable?

File: internal/nodejs-action/generate-layout.sh
Comment (laurentsimon) [inline internal/nodejs-action/generate-layout.sh:4]: thoughts on using TS for this code? Maybe as a follow-up?

File: internal/nodejs-action/action.yml
Comment (asraa) [inline internal/nodejs-action/action.yml:66]: In case someone runs with a matrix strategy, then I believe multiple artifacts would be uploaded to the name `package.tgz`. The `file-path` is likely package-specific and run-specific, so I don't think you'd have to worry about adding an RNG in the path. 

But, using a hard-coded fixed name for the aritfact may be overwritten by matrix or concurrent runs by the same caller

Comment (laurentsimon) [reply internal/nodejs-action/action.yml:66]: I think I agree with Asra. The name is important and needs to be unique across concurrent runs.

File: internal/nodejs-action/generate-layout.sh
Comment (ianlewis) [reply internal/nodejs-action/generate-layout.sh:18]: The name from the output of `npm pack` only contains the `<scope>/<package name>`. It's not a purl. The comment above was intended as a general comment about why I'm parsing out the parts of the name, and url encoding it etc.

Maybe I can add more comments to make that more clear.

File: internal/nodejs-action/generate-layout.sh
Comment (ianlewis) [reply internal/nodejs-action/generate-layout.sh:16]: The version in a purl could be any kind of string. I'm not sure what you're getting at here? Maybe you could elaborate?

File: internal/nodejs-action/generate-layout.sh
Comment (ianlewis) [reply internal/nodejs-action/generate-layout.sh:4]: Yeah, I'd like to use TS as much as possible.

It's just easier to use Bash since to use TS I'd need to write yet another GitHub action with all the boilerplate (`action.yml`, `package.json`, `Makefile` etc.)

I could use [`github-script`](https://github.com/actions/github-script) but that suffers from a lot of the same problems as using bash (hard to test, lint etc.)

I added #1849 to track.

File: internal/nodejs-action/generate-layout.sh
Comment (asraa) [inline internal/nodejs-action/generate-layout.sh:51]: what if it's just the packages `<name>.intoto`?

File: internal/nodejs-action/generate-layout.sh
Comment (laurentsimon) [inline internal/nodejs-action/generate-layout.sh:27]: maybe add a comment that version may be a commit, or a string provided by user

**Review by laurentsimon — APPROVED** [review]

File: .github/workflows/builder_nodejs_slsa3.yml
Comment (laurentsimon) [reply .github/workflows/builder_nodejs_slsa3.yml:120]: unresolving since I don't see the changes and this is an important one

File: internal/nodejs-action/action.yml
Comment (laurentsimon) [inline internal/nodejs-action/action.yml:26]: should this be called provenance-name or attestation-name?

File: internal/nodejs-action/generate-layout.sh
Comment (laurentsimon) [reply internal/nodejs-action/generate-layout.sh:16]: the publish attestation has a version, the package.json has one too; so I wanted to have some clarification about the assumption we're making on these fields. Could go in a .md explaining the design and assumptions end-to-end.

File: .github/workflows/builder_nodejs_slsa3.yml
Comment (ianlewis) [reply .github/workflows/builder_nodejs_slsa3.yml]: Sorry. I'm in the process of pushing the changes and writing the responses and GitHub doesn't allow you to resolve comments as part of a review.

I pinned the digest and updated it so that it fetches the digest to checkout with --depth 1

File: internal/nodejs-action/generate-layout.sh
Comment (ianlewis) [reply internal/nodejs-action/generate-layout.sh]: I updated this to use the package `filename` included in the output of `npm pack --json`. I can't just use the package name since it could be something like `@ianlewis/mypackage` if it's a scoped package.

Right now though, this is only used internally in the workflow as the provenance is uploaded as json metadata by `npm publish` so this file isn't actually stored anywhere.

File: internal/nodejs-action/action.yml
Comment (ianlewis) [reply internal/nodejs-action/action.yml:26]: I think `attestation-name` is more accurate since it's the attestation generated by `generate-attestations` and signed by `sign-attestations`.

This is only really used internally by the workflow so it shouldn't be exposed to users anyway.

File: internal/nodejs-action/generate-layout.sh
Comment (ianlewis) [reply internal/nodejs-action/generate-layout.sh:34]: Done.

File: internal/nodejs-action/action.yml
Comment (ianlewis) [reply internal/nodejs-action/action.yml]: I added an rng value to the uploaded package artifact name.

File: internal/nodejs-action/generate-layout.sh
Comment (ianlewis) [reply internal/nodejs-action/generate-layout.sh:21]: Ok, AFAIK package versions aren't really validated by npm in any way. They could be any kind of string. I think that the slsa-verifier expects versions to be in semver format and that's perhaps fine, though I don't think our builder should make any assumptions about the version.

**Comment by ianlewis** [comment]
@laurentsimon @asraa I made some more changes to address the comments and tested it on my test repo. I think it should be good to go but wanted to give you a chance to take one more pass before merging.

File: .github/workflows/builder_nodejs_slsa3.yml
Comment (laurentsimon) [inline .github/workflows/builder_nodejs_slsa3.yml:95]: so this would mean we'll have to expose the action `secure-download-artifact` to TRWs in the future. Can we instead download insecurely and use the slsa-verifier to verify the artifact is the correct one?

What we'll need is:
1. BYOB support inn slsa-verifier
2. Support for `--source-digest` in slsa-verifier, to prevent someone from replacing the genuine artifact with an older artifact

Wdut?

File: internal/nodejs-action/action.yml
Comment (laurentsimon) [inline internal/nodejs-action/action.yml:80]: so we'll need to expose this function to TRWs, it seems.
It's an interesting and very important use case, I added it to the BYOB doc list https://github.com/slsa-framework/slsa-github-generator/issues/1574#issuecomment-1405878671

File: .github/workflows/builder_nodejs_slsa3.yml
Comment (asraa) [reply .github/workflows/builder_nodejs_slsa3.yml:95]: > I think provenance should also be securely downloaded.

This finally only made sense to me since the TRW process should not be influenced by the maintainer, and we need to avoid the potential provenance confusion Laurent mentioned above.

I expect in this case we are only creating 1 attestation file, but the delegator handles a folder of attestations inc ase the layout specified more. The upload:
https://github.com/slsa-framework/slsa-github-generator/blob/ab4b569619e185570320838296f57451d79d3547/.github/workflows/delegator_generic_slsa3.yml#L257

I think it should be fairly easy to support folders in secure upload/download, if we tar the folders before and after download to checksum

Comment (laurentsimon) [reply .github/workflows/builder_nodejs_slsa3.yml:95]: In fact, I think the provenance is the only file that needs to be securely downloaded. The artifact verification can then be done via slsa-verifier. It would fail if the user changed the artifact. This would probably also simplify the Action. So it would look like this:

1. TRW MUST securely download the attestation we return
2. TRW MUST slsa-verifier with `--source-uri=$GITHUB_REPOSITORY --build-id <their-builder-id>` before uploading artifact

Comment (laurentsimon) [reply .github/workflows/builder_nodejs_slsa3.yml:95]: > I think it should be fairly easy to support folders in secure upload/download, if we tar the folders before and after download to checksum

That sounds like a good idea. Do we have an issue for it already?

Comment (laurentsimon) [reply .github/workflows/builder_nodejs_slsa3.yml:95]: Created https://github.com/slsa-framework/slsa-github-generator/issues/1863.

File: internal/nodejs-action/action.yml
Comment (ianlewis) [reply internal/nodejs-action/action.yml:80]: I was thinking a bit about this. I think we should think more about this. I'm not sure we need to provide a lot of "utilities" to TRW authors or we should provide better ways to pass out info about the artifacts & attestations that are uploaded by the internal action.

File: .github/workflows/builder_nodejs_slsa3.yml
Comment (ianlewis) [reply .github/workflows/builder_nodejs_slsa3.yml:95]: Thanks. I think this is a good approach. I think we can address this on a follow up PR.

File: internal/nodejs-action/action.yml
Comment (ianlewis) [reply internal/nodejs-action/action.yml:80]: I think #1863 covers this pretty well and we won't really need this if we implement that idea.

**Pull request merged**
